website/blog/2024-11-27-test-smarter-part-2.md

```diff
@@ -0,0 +1,123 @@
+- Intermediate models may perform a first set of joins or aggregations to reduce complexity in a final mart.
+    - Add simple anomaly tests to verify the behavior of your sets of joins and aggregations. This may look like:
+        - An [accepted_values](https://docs.getdbt.com/reference/resource-properties/data-tests#accepted_values) test on a newly calculated categorical column.
+        - A [mutually_exclusive_ranges](https://docs.getdbt.com/reference/resource-properties/data-tests#accepted_values) test on two columns whose values behave in relation to one another (ex: asserting age ranges do not overlap).
+        - A [not_constant](https://docs.getdbt.com/reference/resource-properties/data-tests#accepted_values) test on a column whose value should be continually changing (ex: page view counts on website analytics).
+- Intermediate models may isolate complex operations.
+    - The anomaly tests we list above may suffice here.
+    - You might also consider [unit testing](https://docs.getdbt.com/docs/build/unit-tests) any particularly complex pieces of SQL logic.
+
+### Marts
+
+Marts layer testing will follow the same hygiene-or-anomaly pattern as staging and intermediate. Similar to your intermediate layer, you should focus your testing on net-new columns in your marts layer. This might look like:
+
+- Unit tests: validate especially complex transformation logic. For example:
+    - Calculating dates in a way that feeds into forecasting.
+    - Customer segmentation logic, especially logic that has a lot of CASE-WHEN statements.
+- Primary key tests: focus on where where your mart's granularity has changed from its staging/intermediate inputs.
+    - Similar to the intermediate models above, you may also want to add primary key tests to models whose grain hasn’t changed, but have been enriched with other data. Primary key tests here communicate your intent.
+- Business focused anomaly tests: focus on *new* calculated fields, such as:
+    - Singular tests on high-priority, high-impact tables where you have a specific problem you want forewarning about.
+        - This might be something like fuzzy matching logic to detect when the same person is making multiple emails to extend a free trial beyond its acceptable end date.
+    - A test for calculated numerical fields that shouldn’t vary by more than certain percentage in a week.
+    - A calculated ledger table that follows certain business rules, i.e. today’s running total of spend must always be greater than yesterday’s.
+
+### CI/CD
+
+All of the testing you’ve applied in your different layers is the manual work of constructing your framework. CI/CD is where it gets automated. 
+
+You should run a [slim CI](https://docs.getdbt.com/best-practices/best-practice-workflows#run-only-modified-models-to-test-changes-slim-ci) by default, unless there’s good reason to build everything in your project. 
```

Comment:
```suggestion
You should run a [slim CI](/best-practices/best-practice-workflows#run-only-modified-models-to-test-changes-slim-ci) to optimize your resource consumption. 
```

SLIM CI 4 LYF

Reply:
you bet your bottom dollar! yes let's commit this.